Testing evaluating requirements and hints eagerly

kinow opened the pull request:
WIP, similar to #1566, with a tentative fix for the infamous #1330 :cowboy_hat_face: 

Letting CI run to see what breaks after the fix, before adding tests or making further changes.

kinow commented:
Currently failing:

`test_udocker_should_display_memory_usage`

```
2022-08-06T04:27:08.2095534Z Test [182/197] dynamic_resreq_wf_optional_file_default: Within a workflow, test accessing the size attribute of an optional input File as part of a CommandLineTool's ResourceRequirement calculation. The CommandLineTool input has a default value (a local file) and the workflow nor the workflow step does not provide any value for this input.
2022-08-06T04:27:08.2096211Z 
2022-08-06T04:27:09.2255639Z Test [183/197] dynamic_resreq_wf_optional_file_step_default: Within a workflow, test accessing the size attribute of an optional input File as part of a CommandLineTool's ResourceRequirement calculation. The workflow step provides a default value (a local file) for this input and the workflow itself does not provide any value for this input.
2022-08-06T04:27:09.2257721Z 
2022-08-06T04:27:10.5414561Z Test 182 failed: /home/runner/work/cwltool/cwltool/common-workflow-language-main/cwltool_with_cov3 --parallel --podman --outdir=/tmp/tmp8mtpvy82 --quiet v1.0/dynresreq-workflow-tooldefault.cwl v1.0/empty.json
2022-08-06T04:27:10.5415873Z Within a workflow, test accessing the size attribute of an optional input File as part of a CommandLineTool's ResourceRequirement calculation. The CommandLineTool input has a default value (a local file) and the workflow nor the workflow step does not provide any value for this input.
2022-08-06T04:27:10.5416546Z Returned non-zero
2022-08-06T04:27:10.5416963Z [1;30mWARNING[0m [33mWorkflow checker warning:
2022-08-06T04:27:10.5417553Z v1.0/dynresreq-workflow-tooldefault.cwl:6:3:  Source 'special_file' of type ["null", "File"] may be
2022-08-06T04:27:10.5418001Z                                               incompatible
2022-08-06T04:27:10.5418537Z v1.0/dynresreq-workflow-tooldefault.cwl:15:7:   with sink 'special_file' of type "File"[0m
2022-08-06T04:27:10.5419362Z Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.
2022-08-06T04:27:10.5419908Z Traceback (most recent call last):
2022-08-06T04:27:10.5427651Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 308, in do_eval
2022-08-06T04:27:10.5444618Z     return interpolate(
2022-08-06T04:27:10.5445633Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 214, in interpolate
2022-08-06T04:27:10.5446092Z     e = evaluator(
2022-08-06T04:27:10.5446726Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 163, in evaluator
2022-08-06T04:27:10.5447200Z     raise JavascriptException(
2022-08-06T04:27:10.5448037Z cwl_utils.errors.JavascriptException: Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.
2022-08-06T04:27:10.5448821Z [1;30mERROR[0m [31mUnhandled error, try again with --debug for more information:
2022-08-06T04:27:10.5449200Z   Expression evaluation error:
2022-08-06T04:27:10.5449940Z Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.[0m
2022-08-06T04:27:10.5450382Z 
2022-08-06T04:27:10.5451470Z Test [184/197] dynamic_resreq_wf_optional_file_wf_default: Within a workflow, test accessing the size attribute of an optional input File as part of a CommandLineTool's ResourceRequirement calculation. The workflow itelf provides a default value (a local file) for this input.
2022-08-06T04:27:10.5452724Z 
2022-08-06T04:27:11.5594389Z Test 183 failed: /home/runner/work/cwltool/cwltool/common-workflow-language-main/cwltool_with_cov3 --parallel --podman --outdir=/tmp/tmp7v61oixp --quiet v1.0/dynresreq-workflow-stepdefault.cwl v1.0/empty.json
2022-08-06T04:27:11.5595877Z Within a workflow, test accessing the size attribute of an optional input File as part of a CommandLineTool's ResourceRequirement calculation. The workflow step provides a default value (a local file) for this input and the workflow itself does not provide any value for this input.
2022-08-06T04:27:11.5596706Z Returned non-zero
2022-08-06T04:27:11.5597545Z [1;30mWARNING[0m [33mWorkflow checker warning:
2022-08-06T04:27:11.5646518Z v1.0/dynresreq-workflow-stepdefault.cwl:6:3:  Source 'special_file' of type ["null", "File"] may be
2022-08-06T04:27:11.5646899Z                                               incompatible
2022-08-06T04:27:11.5647365Z v1.0/dynresreq-workflow-stepdefault.cwl:15:7:   with sink 'special_file' of type "File"[0m
2022-08-06T04:27:11.5648126Z Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.
2022-08-06T04:27:11.5648596Z Traceback (most recent call last):
2022-08-06T04:27:11.5649242Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 308, in do_eval
2022-08-06T04:27:11.5649603Z     return interpolate(
2022-08-06T04:27:11.5650149Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 214, in interpolate
2022-08-06T04:27:11.5650493Z     e = evaluator(
2022-08-06T04:27:11.5651008Z   File "/home/runner/work/cwltool/cwltool/cwltool-venv3/lib/python3.9/site-packages/cwl_utils/expression.py", line 163, in evaluator
2022-08-06T04:27:11.5651396Z     raise JavascriptException(
2022-08-06T04:27:11.5652146Z cwl_utils.errors.JavascriptException: Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.
2022-08-06T04:27:11.5652834Z [1;30mERROR[0m [31mUnhandled error, try again with --debug for more information:
2022-08-06T04:27:11.5653153Z   Expression evaluation error:
2022-08-06T04:27:11.5654025Z Syntax error in parameter reference 'inputs.special_file.size': inputs does not contain key 'special_file'. This could be due to using Javascript code without specifying InlineJavascriptRequirement.[0m
```

kinow commented:
Conformance tests fixed locally, mypy and linter passing too. Pushed a new commit, waiting for CI. I think mypy failed on GH, but I'm hoping it's just a glitch (otherwise my environment might be using an older version?)

kinow reviewed (commented):
Unit & conformance tests passing :tada: 

Mypy passes locally, but for some reason is failing on GH (maybe some usage limits on the server side?)

```
    File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/subprocess.py", line 311, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['docker', 'pull', 'docker.io/ubuntu']' returned non-zero exit status 1.
  ERROR    cwltool:test_relocate.py:17 Workflow error, try again with --debug for more information:
  Docker is not available for this tool, try --no-container to disable Docker, or install a user space Docker replacement like uDocker with --user-space-docker-cmd.: Command '['docker', 'pull', 'docker.io/ubuntu']' returned non-zero exit status 1.
```

Will try to add the tests from #1566, but I believe it's ready for a review (even though I'm not sure if this is the best fix :+1: )

I tested it with a workflow `1330.cwl`:

```cwl
cwlVersion: v1.2
class: Workflow

hints:
  InlineJavascriptRequirement: {}
  ResourceRequirement:
    coresMax: $(inputs.threads_max)

inputs:
  threads_max:
    type: int
    default: 4

steps:
  one:
    in: []
    run:
      class: CommandLineTool
      inputs:
        other_input:
          type: int
          default: 8
      baseCommand: echo
      arguments: [ $(runtime.cores) ]
      outputs: []
    out: []

outputs: []

```

-Bruno

kinow commented:
From today's APAC-EMEA:

1. Does it work with command-line tool, expression tool? Operation is OK too?
2. Need to add to conformance tests.
3. Need to update specification (perhaps).
4. @tom-tan & @manabuishii think it's going in the right direction too :tada:

kinow commented:
>Does it work with command-line tool, expression tool? Operation is OK too?

It does! :tada: 

I thought I would have to modify `default_make_tool` in other modules (for commandline tool and expression tool), but looks like that function is called for making the tool in the `main.py` for any of the process types we have.

I tested with the following (possible good candidates for conformance tests):

*1330-wf.cwl*
```cwl
cwlVersion: v1.2
class: Workflow

hints:
  ResourceRequirement:
    coresMax: $(inputs.threads_max)

inputs:
  threads_max:
    type: int
    default: 4

steps:
  one:
    in: []
    run:
      class: CommandLineTool
      inputs:
        other_input:
          type: int
          default: 8
      baseCommand: echo
      arguments: [ $(runtime.cores) ]
      outputs: []
    out: []

outputs: []
```

*1330-cmdtool.cwl*
```cwl
cwlVersion: v1.2
class: CommandLineTool

hints:
  ResourceRequirement:
    coresMax: $(inputs.threads_max)

inputs:
  threads_max:
    type: int
    default: 4

outputs: []

arguments: ['echo', $(runtime.cores)]
```

*1330-expr.cwl*
```cwl
cwlVersion: v1.2
class: ExpressionTool

requirements:
  InlineJavascriptRequirement: {}

hints:
  ResourceRequirement:
    coresMax: $(inputs.threads_max)

inputs:
  threads_max:
    type: int
    default: 4

outputs:
  out: string

expression: |
  ${ return {"out": runtime.cores }}
```

Using the default job order, and providing `--threads_max 1`.

-Bruno

mr-c reviewed (changes requested):
Great to see CI passing!

Please add local copies of the suggested conformance tests that did not previously pass, and use pytest to confirm that they now pass. 

I would like @tetron to review this before merging

kinow commented:
> Great to see CI passing!

:tada:

> Please add local copies of the suggested conformance tests that did not previously pass, and use pytest to confirm that they now pass.

Done!

> I would like @tetron to review this before merging

:+1:  thanks @mr-c !

kinow commented:
@mr-c I tried `make cleanup` and it fixed a lot of files locally. CI appears to be running `black` for the diff, but codespell and other tools for all files?

```
tests/wf/workreuse.cwl tests/wf/wrong_cwlVersion.cwl tests/with_doc.cwl tests/without_doc.cwl tox.ini
  FIXED: CODE_OF_CONDUCT.md
  FIXED: Makefile
  FIXED: cwltool/provenance.py
  FIXED: cwltool/provenance_constants.py
  FIXED: cwltool/provenance_profile.py
  FIXED: cwltool/run_job.py
  FIXED: tests/bundle-context.jsonld
  FIXED: tests/test_provenance.py
  CWLProv.rst:47: ro ==> to, row, rob, rod, roe, rot
  CWLProv.rst:73: ro ==> to, row, rob, rod, roe, rot
  CWLProv.rst:73: ro ==> to, row, rob, rod, roe, rot
  cwltool/docker.py:371: ro ==> to, row, rob, rod, roe, rot
  cwltool/executors.py:219: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/main.py:707: ro ==> to, row, rob, rod, roe, rot
  cwltool/main.py:713: ro ==> to, row, rob, rod, roe, rot
  cwltool/main.py:714: ro ==> to, row, rob, rod, roe, rot
  cwltool/main.py:714: ro ==> to, row, rob, rod, roe, rot
  cwltool/main.py:1444: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:306: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:314: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:722: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance.py:732: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:776: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:785: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:878: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:898: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:933: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:953: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance.py:955: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_constants.py:31: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_constants.py:31: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance_constants.py:31: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance_profile.py:47: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:73: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:315: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:326: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:414: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:417: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance_profile.py:420: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:463: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance_profile.py:463: ro ==> to, row, rob, rod, roe, rot
  cwltool/provenance_profile.py:464: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:468: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:482: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:667: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/provenance_profile.py:749: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  cwltool/singularity.py:323: ro ==> to, row, rob, rod, roe, rot
  cwltool/udocker.py:17: ro ==> to, row, rob, rod, roe, rot
  cwltool/workflow.py:91: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/bundle-context.jsonld:11: ro ==> to, row, rob, rod, roe, rot
  tests/bundle-context.jsonld:11: ro ==> to, row, rob, rod, roe, rot
  tests/bundle-context.jsonld:113: ro ==> to, row, rob, rod, roe, rot
  tests/test_provenance.py:26: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:26: ro ==> to, row, rob, rod, roe, rot
  tests/test_provenance.py:303: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:334: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:528: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:542: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:543: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:545: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  tests/test_provenance.py:575: RO ==> TO, ROW, ROB, ROD, ROE, ROT
  make: *** [Makefile:116: codespell] Error 65

```

mr-c commented on tests/test_reqs_hints.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def test_workflow_reqs_are_evaluated_eagerly_default_args() -> None:
+    """Test that a Workflow process will evaluate the requirements eagerly.
+    Uses the default input values.
```
Was this accepted by `make diff_pydocstyle_report`?

kinow replied:
Ah, it was not. Fixed. Thanks @mr-c !

kinow commented:
Rebased, and re-worded the commit that said eager, to say: “Evaluate hints and resources at the workflow level, using the correct inputs.” :+1:

kinow commented:
I was fixing the `cwl-v1.2` repository conformance tests, following @mr-c feedback.

@mr-c, your suggestion to use `EnvVarRequirement` looked interesting! I thought that would simplify things, but it actually pointed to a flaw in the current fix PR.

I am evaluating the `value` of `key: value` of hints & requirements. However, that's not how `cwltool` handles for command line tool. Instead, it handles each requirement in a different way. For `EnvVarRequirement`, it will evaluate—if needed—the values of `envDef`'s keys. For `ToolTimeLimit` it will evaluate the `timelimit` attribute, and so it goes.

I think the proper fix would have to allow the correct `inputs` object to be referenced in any requirement, regardless whether it's a `ResourceRequirement` or any other type.

Moving it back to draft.

mr-c commented:
Review of CWL requirement/hint fields that can contain CWL expressions (inspired by the incomplete list at https://www.commonwl.org/user_guide/13-expressions/index.html )

## CWL v1.0

1. [`InitialWorkDirRequirement`](https://www.commonwl.org/v1.0/Workflow.html#InitialWorkDirRequirement).`listing` single entries
2. [`InitialWorkDirRequirement`](https://www.commonwl.org/v1.0/Workflow.html#InitialWorkDirRequirement).`listing` entries of type [`Dirent`](https://www.commonwl.org/v1.0/Workflow.html#Dirent), fields `entry` and `entryname`
3. [`EnvVarRequirement`](https://www.commonwl.org/v1.0/CommandLineTool.html#EnvVarRequirement) 's [`EnvDef`](https://www.commonwl.org/v1.0/CommandLineTool.html#EnvironmentDef).`envValue`
4. [`ResourceRequirement`](https://www.commonwl.org/v1.0/CommandLineTool.html#ResourceRequirement) fields`coresMin`, `coresMax`, `ramMin`, `ramMax`, `tmpdirMin`, `tmpdirMax`, `outdirMin`, `outdirMax`
5. [`SchemaDefRequirement`](https://www.commonwl.org/v1.0/CommandLineTool.html#SchemaDefRequirement).`types` can can type definitions with `inputBinding`s (defined in [`CommandLineBinding`](https://www.commonwl.org/v1.0/CommandLineTool.html#CommandLineBinding)) that themselves can have CWL expressions in the `valueFrom` field. Unclear if a workflow-level `SchemaDefRequirement` is used, then which `inputs` object is presented to these expressions? Likewise for the use of `format` fields that use CWL expressions.

## CWL v1.1
Same as v1.0 plus

6. [`NetworkAccess`](https://www.commonwl.org/v1.1/Workflow.html#NetworkAccess).`networkAccess`
7. [`SchemaDefRequirement`](https://www.commonwl.org/v1.1/CommandLineTool.html#SchemaDefRequirement).`types` can can type definitions with `inputBinding`s (defined in [`CommandLineBinding`](https://www.commonwl.org/v1.1/CommandLineTool.html#CommandLineBinding)) that themselves can have CWL expressions **also** in the `position` field in addition to the `valueFrom` field. The type definitions could also use a CWL expression in the `secondaryFiles` field.
8. [`ToolTimeLimit`](https://www.commonwl.org/v1.2/Workflow.html#ToolTimeLimit).`timelimit`
9. [`WordReuse`](https://www.commonwl.org/v1.2/CommandLineTool.html#WorkReuse).`enableReuse`

## CWL v1.2
Same as v1.1 plus

10. [`InitialWorkDirRequirement`](https://www.commonwl.org/v1.2/Workflow.html#InitialWorkDirRequirement).`listing` itself

kinow commented:
>Review of CWL requirement/hint fields that can contain CWL expressions (inspired by the incomplete list at https://www.commonwl.org/user_guide/13-expressions/index.html )

Thanks!

It's not hard to review the existing code and confirm these rules/cases match, or if there are any missing. I think the hard part is figuring out a way to re-use that in another context (when evaluating the workflow or workflow steps). Will have a go after the user guide, but happy if anyone beats me to it.
